source/Nevermore/Mapping/IdColumnMapping.cs

```diff
@@ -19,26 +20,92 @@ internal IdColumnMapping(string columnName, Type type, IPropertyHandler handler,
             : base(columnName, type, handler, property)
         { }
 
-        public bool IsIdentity { get; private set; }
+        public IPrimaryKeyHandler? PrimaryKeyHandler { get; private set; }
 
         /// <inheritdoc cref="IIdColumnMappingBuilder"/>
         public IIdColumnMappingBuilder Identity()
+        {
+            ValidateForIdentityUse();
+
+            if (!(PrimaryKeyHandler is null) && !(PrimaryKeyHandler is IIdentityPrimaryKeyHandler))
+                throw new InvalidOperationException($"{nameof(KeyHandler)} has already been set to a non-identity handler.");
+
+            var handlerType = typeof(IdentityPrimaryKeyHandler<>).MakeGenericType(Type);
+            var keyHandler = (IIdentityPrimaryKeyHandler)Activator.CreateInstance(handlerType);
+
+            return KeyHandler(keyHandler);
+        }
+
+        void ValidateForIdentityUse()
         {
             if (!ValidIdentityTypes.Contains(Type))
                 throw new InvalidOperationException($"The type {Type.Name} is not supported for Identity columns. Identity columns must be one of 'short', 'int' or 'long'.");
 
             if (hasCustomPropertyHandler)
                 throw new InvalidOperationException("Unable to configure an Identity Id column with a custom PropertyHandler");
+        }
+
+        public IIdColumnMappingBuilder KeyHandler(IPrimaryKeyHandler primaryKeyHandler)
+        {
+            if (!(PrimaryKeyHandler is null) && Direction == ColumnDirection.FromDatabase && !(primaryKeyHandler is IIdentityPrimaryKeyHandler))
+                throw new InvalidOperationException($"{nameof(KeyHandler)} can only be called with an IIdentityPrimaryKeyHandler, once {nameof(Identity)} has been called.");
 
-            IsIdentity = true;
-            Direction = ColumnDirection.FromDatabase;
+            if (primaryKeyHandler is IIdentityPrimaryKeyHandler)
+            {
+                ValidateForIdentityUse();
+                Direction = ColumnDirection.FromDatabase;
+            }
+
+            PrimaryKeyHandler = primaryKeyHandler;
 
             return this;
         }
 
+        /// <summary>
+        /// Set a function that when given the TableName will return key prefix string.
+        /// </summary>
+        /// <param name="idPrefix">The function to call back to get the prefix.</param>
+        public IIdColumnMappingBuilder Prefix(Func<string, string> idPrefix)
+        {
+            if (!(PrimaryKeyHandler is null) && Direction == ColumnDirection.FromDatabase && PrimaryKeyHandler is IIdentityPrimaryKeyHandler)
```

Comment:
There is a bit of type checking in a few places, I can imagine you tried to remove it but that would break too much of the existing code?